File: lib/domain/cultivation_plan/dtos/public_plan_save_field_datum.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Dtos
      # 公開プラン保存セッションの圃場スナップショット 1 行。
      class PublicPlanSaveFieldDatum
        attr_reader :name, :area, :coordinates

        # @param row [Hash, PublicPlanSaveFieldDatum]
        # @return [PublicPlanSaveFieldDatum, nil]
        def self.from_row(row)
          return row if row.is_a?(PublicPlanSaveFieldDatum)
          return nil unless row.is_a?(Hash)

          new(
            name: fetch_key(row, :name),
            area: fetch_key(row, :area),
            coordinates: Array(fetch_key(row, :coordinates))
          )
        end

        # @param name [String, Numeric, nil]
        # @param area [Numeric, String, nil]
        # @param coordinates [Array]
        def initialize(name:, area:, coordinates: [])
          @name = name.nil? ? nil : name.to_s
          @area = area
          @coordinates = coordinates.freeze
          freeze
        end

        # @return [Hash]
        def to_session_row
          { name: name, area: area, coordinates: coordinates }
        end

        def self.fetch_key(h, key)
          sym = key.to_sym
          str = key.to_s
          h[sym] || h[str]
        end
        private_class_method :fetch_key
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/dtos/public_plan_save_field_datum.rs
//! Ruby: `Domain::CultivationPlan::Dtos::PublicPlanSaveFieldDatum`

use std::collections::BTreeMap;

use serde_json::Value;

/// Ruby: `Domain::CultivationPlan::Dtos::PublicPlanSaveFieldDatum`
#[derive(Debug, Clone, PartialEq)]
pub struct PublicPlanSaveFieldDatum {
    pub name: Option<String>,
    pub area: Option<f64>,
    pub coordinates: Vec<f64>,
}

impl PublicPlanSaveFieldDatum {
    pub fn new(
        name: Option<impl Into<String>>,
        area: Option<f64>,
        coordinates: Vec<f64>,
    ) -> Self {
        Self {
            name: name.map(Into::into),
            area,
            coordinates,
        }
    }

    pub fn from_row(row: &Value) -> Option<Self> {
        let obj = row.as_object()?;
        Some(Self::new(
            Self::map_get(obj, "name").and_then(|v| match v {
                Value::Null => None,
                Value::String(s) => Some(s.clone()),
                Value::Number(n) => Some(n.to_string()),
                _ => None,
            }),
            Self::map_get(obj, "area").and_then(|v| v.as_f64()),
            Self::map_get(obj, "coordinates")
                .map(|v| match v {
                    Value::Array(arr) => arr.iter().filter_map(|x| x.as_f64()).collect(),
                    _ => vec![],
                })
                .unwrap_or_default(),
        ))
    }

    pub fn to_session_row(&self) -> BTreeMap<String, Value> {
        let mut map = BTreeMap::new();
        if let Some(name) = &self.name {
            map.insert("name".into(), Value::String(name.clone()));
        }
        if let Some(area) = self.area {
            map.insert("area".into(), Value::from(area));
        }
        map.insert(
            "coordinates".into(),
            Value::Array(
                self.coordinates
                    .iter()
                    .map(|c| Value::from(*c))
                    .collect(),
            ),
        );
        map
    }

    fn map_get<'a>(h: &'a serde_json::Map<String, Value>, key: &str) -> Option<&'a Value> {
        h.get(key)
    }
}
